### ISSUE TITLE
[release-1.6] Istioctl upgrade failed

### ISSUE THREAD

**Issue by erdun** [istio/istio#23923]
Flow https://preliminary.istio.io/docs/setup/upgrade/

**Bug description**

Flow doc https://preliminary.istio.io/docs/setup/upgrade/ with actions

We run version 1.5.4 current， after run 

```
istioctl install --set revision=canary
```

There will be pod/istiod-xxx with version 1.5.4 and pod/istiod-canary-xxx with version 1.6.0

then run 

```
istioctl upgrade
```
it will be faild

**Expected behavior**

1. Upgrade success
2. What about the canary pod, the document does not guide users what to do next

**Comment by esnible** [comment]
Adding `area/user experience` because the question of what to do next about the canary revision is important and not addressed in 1.6.0 docs.

**Comment by esnible** [comment]
The failure message is:

```
2020-05-18T14:55:22.920278Z	info	Error: failed to read the current Istio version, error: different versions of Istio components found: [ingressgateway pod - istio-ingressgateway-74d686cf48-8b8lr - version: 1.6.0 pilot pod - istiod-74c6d9ccdb-lxvrh - version: 1.5.4 istiod pod - istiod-canary-5d8db667c8-n7f64 - version: 1.6.0]

Error: failed to read the current Istio version, error: different versions of Istio components found: [ingressgateway pod - istio-ingressgateway-74d686cf48-8b8lr - version: 1.6.0 pilot pod - istiod-74c6d9ccdb-lxvrh - version: 1.5.4 istiod pod - istiod-canary-5d8db667c8-n7f64 - version: 1.6.0]
```

**Comment by erdun** [comment]
Yes, there are two pod istiod-xxx and istiod-canary-xxx, but ingress and egress update replace the old, should they run as istiod do

**Comment by esnible** [comment]
@ostromart Would it be reasonable to fix this for 1.6.1?

**Comment by howardjohn** [comment]
Setting revision is not supported on Istio 1.5, only 1.6+

Gateway not having the -canary suffix is intended, you can already create multiple gateways

**Comment by erdun** [comment]
If ingress and gateway are updated by replacing the old ones, you cannot use the label `istio.io/rev` to know which revision the pod belongs to.

And whether to consider adding command `rollback`, we can install pods by revision, but other resources replace the old, that would cause serious errors

**Comment by esnible** [comment]
@ostromart Can you triage this?

**Comment by csdnshyang** [comment]
you can check your istio versin.
what's the client version?

client version: 1.6.12
control plane version: 1.6.12
data plane version: 1.5.4 (3 proxies), 1.6.12 (5 proxies)

**Comment by csdnshyang** [comment]
> Flow https://preliminary.istio.io/docs/setup/upgrade/
> 
> **Bug description**
> 
> Flow doc https://preliminary.istio.io/docs/setup/upgrade/ with actions
> 
> We run version 1.5.4 current， after run
> 
> ```
> istioctl install --set revision=canary
> ```
> 
> There will be pod/istiod-xxx with version 1.5.4 and pod/istiod-canary-xxx with version 1.6.0
> 
> then run
> 
> ```
> istioctl upgrade
> ```
> 
> it will be faild
> 
> **Expected behavior**
> 
> 1. Upgrade success
> 2. What about the canary pod, the document does not guide users what to do next

检查下你的istioctl versin 。
保证istioctl client 版本是1.6.x

**Comment by csdnshyang** [comment]
> Flow https://preliminary.istio.io/docs/setup/upgrade/
> 
> **Bug description**
> 
> Flow doc https://preliminary.istio.io/docs/setup/upgrade/ with actions
> 
> We run version 1.5.4 current， after run
> 
> ```
> istioctl install --set revision=canary
> ```
> 
> There will be pod/istiod-xxx with version 1.5.4 and pod/istiod-canary-xxx with version 1.6.0
> 
> then run
> 
> ```
> istioctl upgrade
> ```
> 
> it will be faild
> 
> **Expected behavior**
> 
> 1. Upgrade success
> 2. What about the canary pod, the document does not guide users what to do next

实在不行 istioctl upgrade --force

**Comment by canyousayyes** [comment]
I hit the same problem. I think the upgrade is either:

1. Canary deployment -> migrate all pods to canary -> delete old istio

1. In-place deployment

But not "Canary + In-place". I think the doc can be written more clearly to avoid such confusion.

---

Anyway in my case, I followed the istio 1.7 upgrade guide (but using istioctl 1.6) to create an istio installation with `--revision=1-6-13`. It seems working fine.

However I found that istioctl 1.6 does not support uninstallation ... now I have 3 istio installation end up in my cluster: the original version, a "canary" revision, a "1-6-13" revision which we're actually using.

Are there anyway I can delete the 2 unused revisions safely? istio 1.7 provides a command `istioctl x uninstall --revision 1-6-5` but I am not sure if it works on older versions.
